### PR TITLE
Update usaco-598.mdx

### PR DESCRIPTION
**Pull request by TruaShamu**
Add java implementation to usaco radio contact.

### CODE REVIEW

File: solutions/usaco-598.mdx [hunk 1]
```diff
@@ -141,4 +141,97 @@ int main() {
...
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class radio {
+public static int[][] fjLocs;
+public static int[][] bLocs;
```
Comment (dongliu0426) [inline solutions/usaco-598.mdx:152]: could you indent this?

File: solutions/usaco-598.mdx [hunk 2]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+		 * FJ to take i steps, and for Bessie to take j steps.
+		 */
+		int[][] dp = new int[N + 1][M + 1];
+
+		//If FJ moves, and Bessie stays still.
+		for (int i = 1; i <= N; i++) {
+			dp[i][0] = dp[i - 1][0] + cost(fjLocs[i], bLocs[0]);
+		}
+
+		//If FJ stays still, and Bessie moves.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:191]:
```suggestion
		// If FJ stays still, and Bessie moves.
```

File: solutions/usaco-598.mdx [hunk 3]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+		for (int j = 1; j <= M; j++) {
+			dp[0][j] = dp[0][j - 1] + cost(fjLocs[0], bLocs[j]);
+		}
+
+		for (int i = 1; i <= N; i++) {
+			for (int j = 1; j <= M; j++) {
+				//The current cost at this cell.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:198]:
```suggestion
				// The current cost at this cell.
```

File: solutions/usaco-598.mdx [hunk 4]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+				int curCost = cost(fjLocs[i], bLocs[j]);
+				//Calculate different costs based on who moves.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:200]:
```suggestion
				// Calculate different costs based on who moves.
```

File: solutions/usaco-598.mdx [hunk 5]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+				dp[i][j] = Math.min(Math.min(bothMove, fjMove), bMove);
+			}
+		}
+
+		pw.println(dp[N][M]);
+		pw.close();
+
+	}
+
+	//Fill the location arrays for FJ and Bessie.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:213]:
```suggestion
	// Fill the location arrays for FJ and Bessie.
```

File: solutions/usaco-598.mdx [hunk 6]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+			//Location chance based on direction.
+			char c = dir.charAt(i - 1);
+			if (c == 'N') { loc[i][1]++; }
+			if (c == 'E') { loc[i][0]++; }
+			if (c == 'S') { loc[i][1]--; }
+			if (c == 'W') { loc[i][0]--; }
+		}
+	}
+
+	//Square of the distance between points a and b.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:228]:
```suggestion
	// Square of the distance between points a and b.
```

File: solutions/usaco-598.mdx [hunk 7]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+		BufferedReader br = new BufferedReader(new FileReader("radio.in"));
+		PrintWriter pw = new PrintWriter(new FileWriter("radio.out"));
+		StringTokenizer st = new StringTokenizer(br.readLine());
+		int N = Integer.parseInt(st.nextToken());
+		int M = Integer.parseInt(st.nextToken());
+
+		fjLocs = new int[N + 1][2]; //FJ's locations.
+		bLocs = new int[M + 1][2]; //Bessie's locations.
+
+		//FJ's original location.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:164]:
```suggestion
		// FJ's original location.
```

File: solutions/usaco-598.mdx [hunk 8]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+public static int[][] bLocs;
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("radio.in"));
+		PrintWriter pw = new PrintWriter(new FileWriter("radio.out"));
+		StringTokenizer st = new StringTokenizer(br.readLine());
+		int N = Integer.parseInt(st.nextToken());
+		int M = Integer.parseInt(st.nextToken());
+
+		fjLocs = new int[N + 1][2]; //FJ's locations.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:161]:
```suggestion
		fjLocs = new int[N + 1][2]; // FJ's locations.
```

File: solutions/usaco-598.mdx [hunk 9]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+		bLocs = new int[M + 1][2]; //Bessie's locations.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:162]:
```suggestion
		bLocs = new int[M + 1][2]; // Bessie's locations.
```

File: solutions/usaco-598.mdx (hunk 7 above)
Comment (nwatx) [inline solutions/usaco-598.mdx:164]:
```suggestion
		// FJ's original location.
```

File: solutions/usaco-598.mdx [hunk 10]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+
+		//FJ's original location.
+		st = new StringTokenizer(br.readLine());
+		fjLocs[0][0] = Integer.parseInt(st.nextToken());
+		fjLocs[0][1] = Integer.parseInt(st.nextToken());
+
+		//Bessie's original location.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:169]:
```suggestion
		// Bessie's original location.
```

File: solutions/usaco-598.mdx [hunk 11]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+		st = new StringTokenizer(br.readLine());
+		bLocs[0][0] = Integer.parseInt(st.nextToken());
+		bLocs[0][1] = Integer.parseInt(st.nextToken());
+
+		//Find locations that the two visit.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:174]:
```suggestion
		// Find locations that the two visit.
```

**Review by nwatx — CHANGES_REQUESTED** [review]

**Review by dongliu0426 — APPROVED** [review]

File: solutions/usaco-598.mdx [hunk 12]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+
+	}
+
+	// Fill the location arrays for FJ and Bessie.
+	public static void fill(int[][] loc, String dir) {
+		for (int i = 1; i <= dir.length(); i++) {
+			loc[i][0] = loc[i - 1][0];
+			loc[i][1] = loc[i - 1][1];
+
+			//Location chance based on direction.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:219]:
```suggestion
			// Location chance based on direction.
```
missed one :p

**Review by nwatx — CHANGES_REQUESTED** [review]

File: solutions/usaco-598.mdx [hunk 13]
```diff
@@ -141,4 +141,97 @@ int main() {
...
+		fill(fjLocs, fjPath);
+		fill(bLocs, bPath);
+
+		/*
+		 * dp[i][j] is the minimum cost for
+		 * FJ to take i steps, and for Bessie to take j steps.
+		 */
+		int[][] dp = new int[N + 1][M + 1];
+
+		//If FJ moves, and Bessie stays still.
```
Comment (nwatx) [inline solutions/usaco-598.mdx:186]:
```suggestion
		// If FJ moves, and Bessie stays still.
```

**Review by nwatx — CHANGES_REQUESTED** [review]

**Review by nwatx — APPROVED** [review]

**Pull request merged**